[WIP] push_notifications: Use uuids to identify users to the bouncer.

mateuszmandera opened the pull request:
Fixes #18017

Remaining TODOs:
1. Generate UserPushNotificationIdentities when importing a realm. - blocker, but easy to do.
2. The tests currently test both id-based and uuid-based requests to the bouncer but in a chaotic, inconsistent way. The tests were patched up to keep their current form, with necessary adjustments made. They'll need to be investigated to be, probably, made more consistent - e.g. tests that manually create requests to send to the bouncer still make id-based requests, while tests that use higher-level functions end up making requests with the new format, using uuids. Perhaps all these tests should be switched to using the new format + some class for testing the legacy format should be extracted? Not sure. I think the tests do their job right of getting everything tested for now - so perhaps organizing this better belong in a follow-up?
3. Figure out how to deploy this best. There is some risk of the bouncer breaking, if there are bugs here, which makes it quite sensitive - and while we have lots of automated tests, in the end this can't be also verified manually without deploying to production, so we'll probably need a careful monitoring and reversal strategy? Perhaps we want to extract a smaller commit with *only* the changes to the bouncer API - so that it can be easily deployed, tested that it maintains back-compatibility with currently-existing self-hosted server, and easily reverted if needed?

timabbott commented on zilencer/models.py:
```diff
@@ -63,10 +63,11 @@ class RemotePushDeviceToken(AbstractPushDeviceToken):
 
     class Meta:
-        unique_together = ("server", "user_id", "kind", "token")
+        unique_together = [("server", "user_id", "kind", "token"), ("user_uuid", "kind", "token")]
```
I think this new index should have `server` in its list too; while UUIDs are unlikely to overlap between different servers, we want different servers to be treated as completely independent.

timabbott commented on zilencer/views.py:
```diff
@@ -215,22 +234,23 @@ def remote_server_notify_push(
     server = validate_entity(entity)
 
-    user_id = payload["user_id"]
+    user_identity = UserPushIndentityCompat(payload.get("user_id"), payload.get("user_uuid"))
```
This can be for a follow-up PR (certainly a follow-up commit!), but I think it'd be a good for the server to, when receiving such a request, rewrite matching PushDeviceToken entries with just a `user_id` to instead have just the `user_uuid`.  

That way, once a remote Zulip server has upgraded to sending the correct format, it is immediately no longer vulnerable to this issue.

timabbott commented:
This looks great @mateuszmandera!  The test changes do feel messy but I don't think prohibitively so; I think it'd be reasonable to proceed to testing with chat.zulip.org at this point.

Thinking about the data import issue -- is it correct that we don't want to include these UUIDs in a data export?  I suppose probably that's correct as a default, but if one were moving a realm to new hosting infrastructure (preserving the hostname), then exporting/importing these UUIDs would mean that clients don't have to update their push registration.

But I guess that would be pointless since the API key changed anyway.  OK, we should just record this reasoning in a comment in the import code, since it's a bit tricky.

alexmv commented:
> For testing in production, we can likely point chat.zulip.org at staging and use that mechanism to verify that it works correctly both before and after upgrading chat.zulip.org.

My parse is that this change needs to be deployed on both staging and czo, and it involves a database migration -- but because that migration is only adding a table and then filling it, it's safe to run that code in staging and not in production for a bit while testing.

We should make sure we're also observing czo after staging is upgraded but before czo is, to make sure we exercise the old id codepaths.

timabbott commented:
That sounds correct to me.

mateuszmandera commented:
@timabbott

> This can be for a follow-up PR (certainly a follow-up commit!), but I think it'd be a good for the server to, when receiving such a request, rewrite matching PushDeviceToken entries with just a `user_id` to instead have just the `user_uuid`.
> 
> That way, once a remote Zulip server has upgraded to sending the correct format, it is immediately no longer vulnerable to this issue.

Hmm, is the idea that the bouncer, upon receiving a request containing both a `user_id` and `user_uuid`, to `update(user_id=None, user_uuid=user_uuid)` all the old `RemotePushDeviceToken`s with `user_id`? Won't that break notifications via these tokens in the mobile app?

> But I guess that would be pointless since the API key changed anyway. OK, we should just record this reasoning in a comment in the import code, since it's a bit tricky.

Is the point here that the API key change requires all mobile app instances to have to re-log in anyway, so effectively revoking all the old tokens?

alexmv commented:
Rebased to fix the conflict.

alexmv commented on zilencer/views.py:
```diff
@@ -153,23 +154,35 @@ def register_remote_server(
 
+    if not (user_id or user_uuid):
+        raise JsonableError(_("Missing user_id or user_uuid"))
+    if user_id and user_uuid:
+        raise JsonableError(_("Specify only one of user_id or user_uuid"))
```
Why does the underlying UserPushIndentityCompat support having both, when we don't accept both at the interface?

It seems cleaner to me to have UserPushIndentityCompat to be user_id `xor` user_uuid, not `or`.

----

Oh, is the case for taking both that it makes the removal easier for the client?

I guess the added complexity is probably worth the simplicity of the removal being able to make only one request.  It may be worth commenting here and/or on the UserPushIndentityCompat class that we have different models for or vs xor based on where it's being used from.

mateuszmandera replied:
@alexmv The idea here is that:
1. This is the "register" endpoint. On other endpoints, we receive both `user_id` and `user_uuid` value from the remote server - so that we can handle *old* token which were registered with `user_id`. E.g. when we receive "send notification to this user" request, we want to grab `RemotePushToken`s matching either the `user_uuid` OR the `user_id` (or the notification wouldn't be sent to the user's devices that registered before 5.0). 
2. On this "register" endpoint, we don't worry about back-compatibility like that - actually, it would probably be wrong to accept a registration with both `user_id` and `user_uuid` - we want to have registration of just one, clear type, Either UUID (post-5.0 servers) or ID (pre-5.0 servers).

Does that make sense?

alexmv commented on zilencer/views.py:
```diff
@@ -153,23 +154,35 @@ def register_remote_server(
+    if user_id and user_uuid:
+        raise JsonableError(_("Specify only one of user_id or user_uuid"))
+
+    if user_id:
+        kwargs: Dict[str, Union[str, int]] = {"user_id": user_id}
+    else:
+        assert user_uuid
+        kwargs = {"user_uuid": user_uuid}
```
I don't know that I find this more understandable than passing both:
```py3
            RemotePushDeviceToken.objects.create(
                user_id=user_id,
                user_uuid=user_uuid,
                server=server,
                kind=token_kind,
                token=token,
                ios_app_id=ios_app_id,
                # last_updated is to be renamed to date_created.
                last_updated=timezone.now(),
            )
```

I don't think this construction adds any type safety.

timabbott replied:
I think the idea might be that we don't want to record the `user_id` field at all if the other field is available, so that servers immediately get improved privacy of not sharing user IDs with us.

alexmv replied:
That doesn't make sense.  This codepath is on the server, and is after a check for `if user_id and user_uuid: raise JsonableError(_("Specify only one of user_id or user_uuid"))`

timabbott replied:
Ahh, yes, I hadn't clicked into the context and thought this was a different code path.  I agree that what you propose would be an improvement here, since then we're explicitly setting the other values to `None`.

timabbott replied:
Changed this, and also changed the falsey checks to have `is not None`, since otherwise `0` would produce unexpected results.

mateuszmandera replied:
@timabbott Hmm, the idea of the checks is that `0` is  not really a valid value and we should treat it like `None`?

alexmv commented on zerver/models.py:
```diff
@@ -2096,6 +2100,12 @@ class Meta:
         unique_together = (("realm", "name"),)
 
 
+class UserPushNotificationIdentity(models.Model):
+    id: int = models.AutoField(auto_created=True, primary_key=True, verbose_name="ID")
+    user_profile: UserProfile = models.OneToOneField(UserProfile, on_delete=CASCADE, null=False)
+    uuid: UUID = models.UUIDField(default=uuid4, unique=True)
+
+
```
Why are we doing this as a new model, and not a new column on UserProfile?

timabbott replied:
That's a good question. I can see two possible reasons:
* We currently exclude this table when doing a data export/import by default.
* The main downside of adding a column to `UserProfile` is that we fetch the value every time we fetch a `UserProfile` object.

alexmv replied:
Excluding based on table is a valid reason -- though we'll need to add logic to regenerate it, and I'm not sure that's going to less effort than null'ing a field it on export and re-filling a field on import.

I don't know that I buy the cost of making the UserProfile table wider is actually a strong reason.  We don't fetch those rows in bulk, and we already have a _lot_ of crud on them.  A UUID seems far more useful to have in the core table than, say, shipping the `notification_sound` column around on every user fetch.

timabbott replied:
I think the main argument is that this UUID is exclusively accessed when sending/registering for push notifications; whereas `notification_sound` is necessary for e.g. `/register` to display the user's settings. 

The other is that UUIDs compress poorly, and we do compress objects in our caches, but it may not really matter.

It's true that for our primary code paths where we fetch users in bulk, we do things like this:
```
@cache_with_key(realm_user_dicts_cache_key, timeout=3600 * 24 * 7)                                       
def get_realm_user_dicts(realm_id: int) -> List[Dict[str, Any]]:                                         
    return UserProfile.objects.filter(                                                                   
        realm_id=realm_id,                                                                               
    ).values(*realm_user_dict_fields)                                                                    
```

which only fetch specific fields. Our main `UserProfile` object cache does include full objects, but they tend to be fetched only 1-5 at a time, so it may not really matter, other than that changing `UserProfile` as a high-throughput model is a slightly more complex database migration.

I'll add that we have two other high-entropy (aka hard to compress) fields in UserProfile; `password` and `api_key`; the latter we plan to move to a separate table.

timabbott replied:
We settled on a UserProfile field here; marking as resolved.

alexmv commented on zerver/tests/test_push_notifications.py:
```diff
@@ -367,12 +375,14 @@ def test_send_notification_endpoint(self) -> None:
             [
                 "INFO:zilencer.views:"
-                f"Sending mobile push notifications for remote user 6cde5f7a-1f7e-4978-9716-49f69ebfc9fe:{hamlet.id}: "
+                f"Sending mobile push notifications for remote user 6cde5f7a-1f7e-4978-9716-49f69ebfc9fe:id:{hamlet.id}: "
```
These test changes should be in the previous commit, where the behaviour changed, so we keep bisectability.

alexmv replied:
Rebased and re-pushed to split the test changes up.

timabbott commented:
I pushed back here addressing https://github.com/zulip/zulip/pull/21175#discussion_r819990126.

timabbott commented:
I pushed some partial work towards reworking this, but you'll need to finish the rearrangement @mateuszmandera as I'm out of time for this project for today.

mateuszmandera commented:
@timabbott Pushed this with the rework. I've split up the migration into commits a bit, to perhaps make the deployment easier.

@alexmv I've replied to your remaining question in https://github.com/zulip/zulip/pull/21175#discussion_r820103441 - can you check if that sounds right to you?

timabbott commented on zilencer/models.py:
```diff
@@ -63,10 +63,14 @@ class RemotePushDeviceToken(AbstractPushDeviceToken):
+        unique_together = [
+            ("server", "user_id", "kind", "token"),
+            ("server", "user_uuid", "kind", "token"),
+        ]
```
I guess this is implicitly taking advantage of the fact that Django unique indexes will ignore cases where one of the 4 fields, in this case `user_uuid`, is `null`?  I think that's OK, but would probably benefit from a comment explaining that detail.

alexmv replied:
I think this is a feature of Postgres, not of Django?

timabbott replied:
Yeah, I think that's right; but the point is that this index doesn't do what it appears to do to the naive reader with respect to nulls (namely allowing only one (server, null, kind, token) row to exist), and that's what we want, but we should point out that we're taking advantage of this subtle behavior.

mateuszmandera replied:
Added a comment on that here - however, `token` is generally  going be unique on its own I think? So that alone will cause that the `null` aspect won't actually come into play.

timabbott commented on zerver/migrations/0379_userprofile_uuid_backfill.py:
```diff
@@ -0,0 +1,40 @@
+        ).only("id", "uuid"):
+            user_profile.uuid = uuid.uuid4()
+            user_profiles_to_update.append(user_profile)
+        lower_bound += BATCH_SIZE + 1
```
We generally prefer to write this sort of migration where we're adding BATCH_SIZE and making the loops non-inclusive on one side; it's just easier to read that way.

I think one can implement that by making it `id_gt=lower_bound`, and having `lower_bound` start as 0.

alexmv replied:
To be explicit, you also need to adjust so `lower_bound += BATCH_SIZE` not `BATCH_SIZE + 1`, as well.  Otherwise you do 0 - 10000 inclusive in the first batch, then 10002 to 20001, then 20003 to 30002, ...

mateuszmandera replied:
```
dff --git a/zerver/migrations/0379_userprofile_uuid_backfill.py b/zerver/migrations/0379_userprofile_uuid_backfill.py
index 792dc562af..9750f6b862 100644
--- a/zerver/migrations/0379_userprofile_uuid_backfill.py
+++ b/zerver/migrations/0379_userprofile_uuid_backfill.py
@@ -14,16 +14,16 @@ def backfill_user_profile_uuid(apps: StateApps, schema_editor: DatabaseSchemaEdi
         return
 
     BATCH_SIZE = 10000
-    lower_bound = 1
+    lower_bound = 0
 
-    while lower_bound <= max_id:
+    while lower_bound < max_id:
         user_profiles_to_update = []
         for user_profile in UserProfile.objects.filter(
-            id__gte=lower_bound, id__lte=lower_bound + BATCH_SIZE, uuid=None
+            id__gt=lower_bound, id__lte=lower_bound + BATCH_SIZE, uuid=None
         ).only("id", "uuid"):
             user_profile.uuid = uuid.uuid4()
             user_profiles_to_update.append(user_profile)
-        lower_bound += BATCH_SIZE + 1
+        lower_bound += BATCH_SIZE
 
         UserProfile.objects.bulk_update(user_profiles_to_update, ["uuid"])
 
```

Please check if that looks right, these always feels easy to trip up on

alexmv replied:
Looks right to me!

timabbott reviewed (commented):
Posted a couple comments, but otherwise this lgtm.

alexmv commented on zilencer/views.py:
```diff
@@ -153,17 +154,25 @@ def register_remote_server(
     try:
         with transaction.atomic():
             RemotePushDeviceToken.objects.create(
+                # Exactly one of these two user identify fields will be None.
```
```suggestion
                # Exactly one of these two user identity fields will be None.
```

alexmv commented on zerver/models.py:
```diff
@@ -1684,7 +1684,7 @@ class UserProfile(AbstractBaseUser, PermissionsMixin, UserBaseSettings):
     # notifications bouncer that will not have collisions after doing
     # a data export and then import.
-    uuid: Optional[UUID] = models.UUIDField(null=True, default=uuid4, unique=True)
+    uuid: UUID = models.UUIDField(default=uuid4, unique=True)
```
Might be worth a note in the commit message that this is a separate commit to ensure that the backfill in the previous commit runs, so there are no nulls by the time we make this non-optional.

timabbott replied:
Yeah, agreed, that'd be very helpful.

mateuszmandera replied:
True - added a note on that

mateuszmandera commented:
@timabbott  @alexmv  Addressed the discussions above, rebased and added comments on why `register_remote_push_device` doesn't accept user_id+user_uuid and in `UserPushIndentityCompat.filter_q` as well:
```
    if user_id is not None and user_uuid is not None:
        # We don't want "hybrid" registrations with both.
        # Our RemotePushDeviceToken should be either in the new uuid format
        # or the legacy id one.
        raise JsonableError(_("Specify only one of user_id or user_uuid"))
```

```
    def filter_q(self) -> Q:
        """
        This aims to support correctly querying for RemotePushDeviceToken.
        If only one of (user_id, user_uuid) is provided, the situation is trivial,
        If both are provided, we want to query for tokens matching EITHER the
        uuid or the id - because the user may have devices with old registrations,
        so user_id-based, as well as new registration with uuid. Notifications
        naturally should be sent to both.
        """
        if self.user_id is not None and
```

alexmv commented:
Rebased for the conflict I added in 19dfd8e6a7e5b49ad49b29aee5bcda44be252c88.

alexmv commented on zilencer/models.py:
```diff
@@ -63,10 +63,18 @@ class RemotePushDeviceToken(AbstractPushDeviceToken):
     server: RemoteZulipServer = models.ForeignKey(RemoteZulipServer, on_delete=models.CASCADE)
     # The user id on the remote server for this device
-    user_id: int = models.BigIntegerField(db_index=True)
+    user_id: int = models.BigIntegerField(db_index=True, null=True)
```
This `db_index` on `user_id` has literally never been used, because of the unique index which is a better prefix key.  It'd be nice to drop it while we're in here.

alexmv commented on zerver/tests/test_push_notifications.py:
```diff
@@ -76,6 +77,7 @@
     get_display_recipient,
     get_realm,
     get_stream,
+    get_user_profile_by_id,
```
This is unused in this commit

mateuszmandera replied:
@alexmv Moved to the correct ` push_notifications: Use uuids to identify users to the bouncer.` commit

timabbott commented:
Merged, huge thanks for doing this @mateuszmandera!
